src/communication/tlsvchannel.cpp

```diff
@@ -0,0 +1,210 @@
+    }
+
+    mbedtls_ssl_set_bio(&mSsl, this, TLSVChannel::TlsWrite, TLSVChannel::TlsRead, nullptr);
+
+#if defined(MBEDTLS_DEBUG_C)
+    mbedtls_debug_set_threshold(cDebugLevel);
+#endif
+
+    mbedtls_ssl_conf_dbg(&mConf, mDbg, nullptr);
+
+    return mVChannel.Init(xsReadPath, xsWritePath);
+}
+
+aos::Error TLSVChannel::Connect()
+{
+    mbedtls_ssl_session_reset(&mSsl);
+
+    auto err = mVChannel.Connect();
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    int ret;
+    while ((ret = mbedtls_ssl_handshake(&mSsl)) != 0) {
+        if (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
+            return AOS_ERROR_WRAP(ret);
+        }
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error TLSVChannel::Close()
+{
+    return mVChannel.Close();
+}
+
+bool TLSVChannel::IsConnected() const
+{
+    return mVChannel.IsConnected();
+}
+
+aos::Error TLSVChannel::Read(aos::Array<uint8_t>& data, size_t size)
+{
+    auto ret = mbedtls_ssl_read(&mSsl, data.Get(), size);
+    if (ret < 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error TLSVChannel::Write(const aos::Array<uint8_t>& data)
+{
+    auto ret = mbedtls_ssl_write(&mSsl, data.Get(), data.Size());
+    if (ret < 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+aos::RetWithError<mbedtls_svc_key_id_t> TLSVChannel::SetupOpaqueKey(
+    mbedtls_pk_context& pk, const aos::crypto::PrivateKeyItf& privKey)
+{
+    auto statusAddKey = AosPsaAddKey(privKey);
+    if (!statusAddKey.mError.IsNone()) {
+        return statusAddKey;
+    }
+
+    auto ret = mbedtls_pk_setup_opaque(&pk, statusAddKey.mValue);
+    if (ret != 0) {
+        AosPsaRemoveKey(statusAddKey.mValue);
+
+        return aos::RetWithError<mbedtls_svc_key_id_t>(statusAddKey.mValue, AOS_ERROR_WRAP(ret));
+    }
+
+    return aos::RetWithError<mbedtls_svc_key_id_t>(statusAddKey.mValue, aos::ErrorEnum::eNone);
+}
+
+/***********************************************************************************************************************
+ * Static functions
+ **********************************************************************************************************************/
+
+int TLSVChannel::TlsWrite(void* ctx, const unsigned char* buf, size_t len)
+{
+    auto channel = static_cast<TLSVChannel*>(ctx);
+
+    auto err = channel->mVChannel.Write(aos::Array<uint8_t>(buf, len));
+
+    return err == aos::ErrorEnum::eNone ? len : MBEDTLS_ERR_SSL_WANT_WRITE;
+}
+
+int TLSVChannel::TlsRead(void* ctx, unsigned char* buf, size_t len)
+{
+    auto                channel = static_cast<TLSVChannel*>(ctx);
+    aos::Array<uint8_t> data;
```

Comment:
construct data with a pointer to buf and size len
then resize to maximum
then read
and finally set appropriate size

Reply:
Done